pkg/api/customization/secret/validator.go

```diff
@@ -0,0 +1,48 @@
+package secret
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/rancher/norman/api/access"
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types"
+	v1 "github.com/rancher/types/apis/core/v1"
+)
+
+func Validator(request *types.APIContext, schema *types.Schema, data map[string]interface{}) error {
+	if request.Method == http.MethodPost {
+		id := ""
+
+		// extracting project name from data
+		if projectData, ok := data["projectId"].(string); ok {
+			if projectParts := strings.Split(projectData, ":"); len(projectParts) > 1 {
+				id = fmt.Sprintf("%s:%s", projectParts[1], data["name"])
+			}
+		}
+
+		// minimum info needed to use CanDo
+		secretState := map[string]interface{}{
+			"name":        data["name"],
+			"id":          id,
+			"namespaceId": data["namespaceId"],
+		}
+
+		// update is used here to avoid the application of general user permissions for secrets
+		if err := request.AccessControl.CanDo(v1.SecretGroupVersionKind.Group, v1.SecretResource.Name, "update", request, secretState, schema); err != nil {
+			return httperror.NewAPIError(httperror.PermissionDenied, fmt.Sprintf("unable to validate: user does not have permission"))
+		}
+	} else if request.Method == http.MethodPut {
+		var secretState map[string]interface{}
+
+		if err := access.ByID(request, request.Version, request.Type, request.ID, &secretState); err != nil {
+			return httperror.NewAPIError(httperror.NotFound, fmt.Sprintf("unable to validate: %v", err))
```

Comment:
you dont know this is a notFound error here.
Technically, the right thing to do would be to check to see if the error is a not found error and if it is, return NotFound, else return a 500 error with just the raw error message.